kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/optimizations/KonanBCEForLoopBodyTransformer.kt

```diff
@@ -0,0 +1,244 @@
+    }
+
+    private fun checkLastElement(last: IrExpression, loopHeader: ProgressionLoopHeader): BoundsCheckAnalysisResult {
+        var result = BoundsCheckAnalysisResult(false, null)
+        if (last is IrCall) {
+            result = lessThanSize(last)
+            if (last.isGetSizeCall() && !loopHeader.headerInfo.isLastInclusive) {
+                result = BoundsCheckAnalysisResult(true, (last.dispatchReceiver as? IrGetValue)?.symbol)
+            }
+        }
+        return result
+    }
+
+    private fun IrExpression.isProgressionPropertyGetter(propertyName: String) =
+            this is IrCall && symbol.owner.origin == IrDeclarationOrigin.DEFAULT_PROPERTY_ACCESSOR &&
+                    (symbol.signature as? IdSignature.AccessorSignature)?.propertySignature?.asPublic()?.shortName == propertyName &&
+                        dispatchReceiver?.type?.getClass()?.symbol in context.ir.symbols.progressionClasses
+
+    private fun analyzeLoopHeader(loopHeader: ForLoopHeader): BoundsCheckAnalysisResult {
+        analysisResult = BoundsCheckAnalysisResult(false, null)
+        when(loopHeader) {
+            is ProgressionLoopHeader ->
+                when (loopHeader.headerInfo.direction) {
+                    ProgressionDirection.INCREASING -> {
+                        // Analyze first element of progression.
+                        if (!loopHeader.headerInfo.first.compareIntegerNumericConst { it >= 0 }) {
+                            return analysisResult
+                        }
+                        // TODO: variable set to const value. Add constant propagation?
+                        // Analyze last element of progression.
+                        if (loopHeader.headerInfo.last is IrCall) {
+                            val functionCall = (loopHeader.headerInfo.last as IrCall)
+                            // Case of range with step - `for (i in 0..array.size - 1 step n)`.
+                            if (loopHeader.headerInfo.progressionType.getProgressionLastElementFunction == functionCall.symbol) {
+                                val nestedLastVariable = functionCall.getValueArgument(1)
+                                if (nestedLastVariable is IrGetValue && nestedLastVariable.symbol.owner is IrVariable) {
+                                    val nestedLast = (nestedLastVariable.symbol.owner as IrVariable).initializer
+                                    analysisResult = checkLastElement(nestedLast!!, loopHeader)
+                                }
+                            } else {
+                                // Simple progression.
+                                analysisResult = checkLastElement(functionCall, loopHeader)
+                            }
+                        }
+                    }
+                    ProgressionDirection.DECREASING -> {
+                        val valueToCompare = if (loopHeader.headerInfo.isLastInclusive) 0 else -1
```

Comment:
Yes, I know that this can't now happen in lowering, but there is no guarantee that later logic in lowering won't be changed  and K/N team lost this case. I decided to write basic logic that use all information from lowering without consumptions.